### PR TITLE
feat: add support for psbt requests

### LINKED ISSUE #3310: Support PSBTs
Wanted by Gamma and Ordswap

### LINKED ISSUE #3387: Include Taproot address(es) in userData during authentication
To complement the return of the Native SegWit address in the same object per https://github.com/hirosystems/stacks-wallet-web/pull/3247

Perhaps as:

```
{
  btcAddress: {
    p2tr: {
      mainnet: "bc1p8rwmsmm782e7vfae2l2ssplwemqrfa7e3qz56yhfw59tknjx75cs8vlxt2",
      testnet: "tb1pfsm8ku24w9u2aewve23dn3lk4hanc25p7dh8zhcqpfutr5ms38es3653g4"
    }
  }
}
```

### PR DESCRIPTION
**Pull request by fbwoolf**
> Try out this version of the Hiro Wallet - download [extension builds](https://github.com/hirosystems/stacks-wallet-web/actions/runs/4410767308).

This PR implements a new `psbtRequest` method on the `StacksProvider` which allows apps to request that the wallet sign a Partially Signed Bitcoin Transaction (PSBT). There is a supporting PR up in Connect (linked below) with changes there to expose the new request method. The request accepts a required `hex` of the psbt and an optional `signAtIndex` array of indexes and `allowedSighash` array. If an index is not provided then all inputs are signed. A `hex` is returned to the app of the signed PSBT.

A `Bitcoin` tab has been added to the test-app to test this and to provide an example implementation. Currently, the wallet assumes the app is providing witness data in the inputs, so a public key has been provided here thru the auth response for the app to use. Again, refer to the test-app example.

Open to feedback regarding implementation, and happy to change as needed.

PR in Connect: https://github.com/hirosystems/connect/pull/308 (merged/released)

Screen Shot 2023-03-09 at 2 15 19 PM

### CODE REVIEW

**Comment by fbwoolf** [comment]
~~Note: branch will be broken until I install the updated changes from Connect. I just have them working locally atm.~~ Updated with dependency changes from Connect.

**Comment by fbwoolf** [comment]
Todo items for me here:

~~- [ ] Decode hex to show psbt data in the UI so user can verify~~ v2
- [x] Remove `actionType` from request (do not sign and broadcast)

Please comment here if I am missing something.

**Comment by kevinfaveri** [comment]
Would there be a way for not broadcasting the transaction on sign? actually, without be cool to have both signAndBroadcast; and signOnly (without broadcasting)

**Comment by markmhendrickson** [comment]
Feedback from @orenyomtov 

> i see you are using txHex as a parameter name which is confusing - because it is not a bitcoin tx in hex format - but a PSBT in hex format - which is a different thing - it's not crucial, but just semantics

**Comment by fbwoolf** [comment]
> Would there be a way for not broadcasting the transaction on sign? actually, without be cool to have both signAndBroadcast; and signOnly (without broadcasting)

I have removed the `actionType` from the request altogether. We will only sign and return the psbt to the app ...no broadcasting.

**Comment by fbwoolf** [comment]
> 



> 



> Feedback from @orenyomtov
> 
> > i see you are using txHex as a parameter name which is confusing - because it is not a bitcoin tx in hex format - but a PSBT in hex format - which is a different thing - it's not crucial, but just semantics

Yep, I agree here, I thought the same thing over weekend but haven't had much time to work on this again yet. I will change it to just `hex` or `psbtHex`.

File: src/app/components/warning-label.tsx
Comment (kyranjamie) [inline src/app/components/warning-label.tsx]: This should be a component, not inline jsx

File: src/app/pages/psbt-request/components/page-top.tsx
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/page-top.tsx]: I understand copied, but let's name something better

File: src/app/pages/psbt-request/components/psbt-displayer.tsx
Comment (kyranjamie) [inline src/app/pages/psbt-request/components/psbt-displayer.tsx]: fn name

File: src/app/store/psbts/requests.hooks.ts
Comment (kyranjamie) [inline src/app/store/psbts/requests.hooks.ts]: ?? more appropriate

File: src/app/pages/psbt-request/components/psbt-displayer.tsx
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/psbt-displayer.tsx]: Coming back to this bc was in-progress. I added comments for a v2 bc we are just delivering the hex in the UI for v1.

File: src/app/pages/psbt-request/components/page-top.tsx
Comment (fbwoolf) [reply src/app/pages/psbt-request/components/page-top.tsx]: It seems like we need to make these request components more shared components at some point too?

File: src/app/components/warning-label.tsx
Comment (fbwoolf) [reply src/app/components/warning-label.tsx]: Hoping changing this component to use the `Flag` and adding the title didn't break things elsewhere, but will check other spots.

**Comment by fbwoolf** [comment]
@kyranjamie sorry, I think this was more in a draft state before but did some refactoring/renaming if want to review now. Hoping ready to be merged. Updated the `micro-btc-signer` library here and the nec Connect pkgs to make this work have been released/installed as well. All seems to be working.

EDIT: I think there is a lot more we could do here to refactor the requests and share components. I hadn't coded a new request before but there is a lot of code that seems repeated/outdated. If we think it is a priority, I can stop to do it but might be best to make a code health issue for it? Probably another good design issue would be to explore these requests and make them better? @markmhx 

Note: The request option `txHex` was renamed to `hex`.

@mica000 can you review the design bc the header title needed rewording...
Screen Shot 2023-03-09 at 2 15 19 PM

**Comment by kevinfaveri** [comment]
hey folks, why isnt signAtIndex an array also? it should, shouldnt it? In case I wanna sign more than a single input (and dont want to sign all)

**Comment by orenyomtov** [comment]
> hey folks, why isnt signAtIndex an array also? it should, shouldnt it? In case I wanna sign more than a single input (and dont want to sign all)

I 2nd this

**Comment by mica000** [comment]
LGTM!

**Comment by fbwoolf** [comment]
@kyranjamie for tom, it appears I'll need to add a way here to detect if the inputs are segwit or taproot bc when we sign the inputs they have to match? So, I'll need to pass taproot key thru auth too? And, I'll have to detect it somehow before signing?

**Comment by fbwoolf** [comment]
> hey folks, why isnt signAtIndex an array also? it should, shouldnt it? In case I wanna sign more than a single input (and dont want to sign all)

This is easy to change, np. I just didn't realize it was what apps would want. I'll have to make a change in Connect for the `signAtIndex` to accept an array of numbers but an easy fix. Let me know if there is anything else. Thanks for feedback.

**Comment by kyranjamie** [comment]
> @kyranjamie for tom, it appears I'll need to add a way here to detect if the inputs are segwit or taproot bc when we sign the inputs they have to match? So, I'll need to pass taproot key thru auth too? And, I'll have to detect it somehow before signing?

What taproot inputs are needing to be signed? Can you elaborate on that use case? For example, how does an app know which tr address to use? Inscription could be over many.

**Comment by orenyomtov** [comment]
> > hey folks, why isnt signAtIndex an array also? it should, shouldnt it? In case I wanna sign more than a single input (and dont want to sign all)
> 
> 
> 
> This is easy to change, np. I just didn't realize it was what apps would want. I'll have to make a change in Connect for the `signAtIndex` to accept an array of numbers but an easy fix. Let me know if there is anything else. Thanks for feedback.

Can it be an optional field that if omitted, it signs all the inputs it has the key for?

**Comment by fbwoolf** [comment]
Note here, the `signAtIndex` request option has been changed to an array.

Change in Connect: https://github.com/hirosystems/connect/pull/310

**Comment by fbwoolf** [comment]
> 
> Can it be an optional field that if omitted, it signs all the inputs it has the key for?

Yep, it is optional or we sign all.

**Comment by orenyomtov** [comment]
Awesome!

Right now the confirmation dialog shows the PSBT in hex - so it's kind of like blind signing.

I'm sure you have plans to add a beautiful UI to show the PSBT - but in the meantime - it would be great if you could show a JSON representation of the PSBT.

That way it's at least readable by a power-user.

P.S. any timeline you have in mind to release this?

**Comment by markmhendrickson** [comment]
> Right now the confirmation dialog shows the PSBT in hex - so it's kind of like blind signing. I'm sure you have plans to add a beautiful UI to show the PSBT - but in the meantime - it would be great if you could show a JSON representation of the PSBT.

That's indeed our plan, to go from `hex` -> `JSON` -> `carefully crafted display of what's happening` as progressive enhancements here. JSON coming up next.

> P.S. any timeline you have in mind to release this?

Unless @fbwoolf is tracking anything else missing here for a v1, it seems to me that we could release on Monday.

**Comment by crypticmeta** [comment]
waiting eagerly to add this to my site.  This is lit. I tested it with oren's PSBT tx. it worked like a charm.

**Comment by fbwoolf** [comment]
> 
> Unless @fbwoolf is tracking anything else missing here for a v1, it seems to me that we could release on Monday.

Releasing this with the JSON would be an easy change, but decoding it and making it look nice would be for v2. The JSON would list out pretty long for the record bc Uint8Arrays will list out vertically (see image as an example from an early test).

Screen Shot 2023-03-03 at 7 43 36 PM

**Comment by crypticmeta** [comment]
I suppose you can take a look at what unisat does. They display it quite nicely.

File: src/app/store/accounts/blockchain/bitcoin/taproot-account.hooks.ts
Comment (kyranjamie) [inline src/app/store/accounts/blockchain/bitcoin/taproot-account.hooks.ts]: ⚠️ passing account index as address index

**Review by kyranjamie — CHANGES_REQUESTED** [review]

File: src/app/store/accounts/blockchain/bitcoin/taproot-account.hooks.ts
Comment (fbwoolf) [reply src/app/store/accounts/blockchain/bitcoin/taproot-account.hooks.ts]: I believe we can just pass `0` here for now as we are just providing the first address?

Comment (fbwoolf) [reply src/app/store/accounts/blockchain/bitcoin/taproot-account.hooks.ts]: Tested and works with hardcoding `0`, left a comment in the code that it is only temporary until the address api is ready. If there is a better temp solution, let me know or fix before me tom.

**Comment by fbwoolf** [comment]
@markmhx hoping this is ready. Updated the UI with JSON... and the hex in a drawer...

Screen Shot 2023-03-12 at 4 54 47 PM

Screen Shot 2023-03-12 at 4 55 18 PM

Screen Shot 2023-03-12 at 5 01 13 PM

Screen Shot 2023-03-12 at 5 00 06 PM

File: package.json
Comment (fbwoolf) [inline package.json]: Stacks packages need to updated here, and removed from resolutions, once the Connect PR is merged/released.

Comment (fbwoolf) [reply package.json]: https://github.com/hirosystems/connect/pull/310

File: test-app/src/components/bitcoin.tsx
Comment (kyranjamie) [inline test-app/src/components/bitcoin.tsx]: 👀  uses native segwit method

oh, this is because we're trying to spend a native segwit output

Comment (fbwoolf) [reply test-app/src/components/bitcoin.tsx]: I do think I messed up on this and just over looked changing to the taproot pubkey.

Comment (fbwoolf) [reply test-app/src/components/bitcoin.tsx]: Fixed this.

**Comment by fbwoolf** [comment]
@kyranjamie the Connect change has merged/released and I've updated the pkgs here. This needs CR approval at some point.

**Review by edu-stx — APPROVED** [review]

**Comment by fbwoolf** [comment]
@kyranjamie I feel reluctant to merge this w/out an approval from you, but I know @markmhx wants me to get it merged and create a release PR so it is ready for you all in the morning. I am going to merge this, but please feel free to revert or make changes to it off dev in your AM if feedback comes back that we need to make changes or you feel it isn't the right implementation.

File: tests-legacy/integration/settings/settings.spec.ts
Comment (fbwoolf) [inline tests-legacy/integration/settings/settings.spec.ts:1]: These tests were accidentally removed by Edu off dev so adding them back here.

**Comment by igorsyl** [comment]
OMG @fbwoolf You are AMAZING! @radicleart is integrating this with the [sBTC Bridge](https://github.com/trust-Machines/sbtc-bridge/) immediately!

**Comment by markmhendrickson** [comment]
FYI we have some new docs for integrating PSBTs here: https://hirowallet.gitbook.io/developers/apis/sign-transactions/partially-signed-bitcoin-transactions-psbts

**Pull request merged**
